theforeman.org/pipelines/lib/packaging.groovy

```diff
@@ -63,3 +108,319 @@ def repoclosures(repo, releases, version) {
+        } else {
+            sh """
+              # Download sources
+              wget https://downloads.theforeman.org/${project}/${project}-${package_version}.tar.bz2 https://downloads.theforeman.org/${project}/${project}-${package_version}.tar.bz2.sig
+
+              mv ${project}-${package_version}.tar.bz2 ${project}_${package_version}.orig.tar.bz2
+            """
+        }
+
+        sh """
+            # Unpack
+            tar xvjf ${project}_${package_version}.orig.tar.bz2
+            if [[ -d ${project}-${package_version}-develop ]] ; then
+              mv ${project}-${package_version}-develop ${project}-${package_version}
+            fi
+
+            # Bring in the debian packaging files
+            cp -r ../${project} ./${project}-${package_version}/debian
+        """
+
+        dir("${project}-${package_version}") {
+            if (pull_request) {
+                add_debian_changelog(os, package_version, repoowner, last_commit)
+                sh "mv ../${project}_${package_version}.orig.tar.bz2 ../${project}_9999.orig.tar.bz2"
+            }
+        }
+    }
+
+    return "${build_dir}/${project}-${package_version}"
+}
+
+def setup_sources_dependency(project, os, version, repoowner, pull_request = false) {
+    def package_version = sh(returnStdout: true, script: "head -n1 dependencies/${os}/${project}/changelog|awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-|cut -d: -f2").trim()
+    def build_dir = deb_build_dir('dependency', project, os)
+    def package_dir
+
+    sh "mkdir -p ${build_dir}"
+
+    dir(build_dir) {
+        sh """
+            # Import variables from the project, allowing it to override behaviour
+            if [ -e ../${project}/build_vars.sh ]; then
+              . ../${project}/build_vars.sh
+            fi
+
+            if [[ \$BUILD_TYPE == "python" ]]; then
+              pip download --no-deps --no-binary=:all: "${project}==${package_version}"
+              mkdir "${project}-${package_version}"
+              tar -x -C "${project}-${package_version}" --strip-components=1 -f "${project}-${package_version}.tar.gz"
+              mv "${project}-${package_version}.tar.gz" "${project}_${package_version}.orig.tar.gz"
+              cp -r ../${project} "${project}-${package_version}/debian"
+            elif [[ \$BUILD_TYPE == "gem" ]]; then
+              gem fetch ${project} -v "=${package_version}"
+              ../../gem2deb ${project}-${package_version}.gem --debian-subdir ../${project} --only-source-dir
+            else
+              echo "Unsupported build type: \${BUILD_TYPE}"
+              exit 1
+            fi
+        """
+
+        package_dir = sh(returnStdout: true, script: "find -maxdepth 1 -type d -not -name '.'").trim()
+    }
+
+    dir("${build_dir}/${package_dir}") {
+        if (pull_request) {
+            last_commit = sh(reutrnStdout: true, script: "git rev-list HEAD|/usr/bin/head -n 1").trim()
```

Comment:
we already have `archive_git_hash` in `git.groovy`, which does `git rev-parse ${ref}` (and that's equal to `rev-list and then getting the first entry)